benchexec/container.py

```diff
@@ -834,6 +928,32 @@ def drop_capabilities(keep=[]):
         ctypes.byref(libc.CapHeader(version=libc.LINUX_CAPABILITY_VERSION_3, pid=0)),
         ctypes.byref(capdata),
     )
+    libc.prctl(libc.PR_CAP_AMBIENT, libc.PR_CAP_AMBIENT_CLEAR_ALL, 0, 0, 0)
+
+
+def cap_permitted_to_ambient():
+    """
+    Python version of util-linux/lib/caputils.c: cap_permitted_to_ambient()
+    """
+
+    header = libc.CapHeader(libc.LINUX_CAPABILITY_VERSION_3, 0)
+    data = (libc.CapData * libc.LINUX_CAPABILITY_U32S_3)()
+
+    libc.capget(header, data)
+
+    data[0].inheritable = data[0].permitted
+    data[1].inheritable = data[1].permitted
+
+    libc.capset(header, data)
+
+    effective = (data[1].effective << 32) | data[0].effective
+    cap_last_cap = int(util.try_read_file("/proc/sys/kernel/cap_last_cap") or "0")
+    for cap in range(cap_last_cap + 1):
+        if cap > cap_last_cap:
+            continue
```

Comment:
This is the same check as `cap_valid()`, but using the runtime value for the last valid cap.

Reply:
But the loop only iterates until `cap_last_cap` anyway?

Reply:
Sorry about that. I just reviewed the `cap_permitted_to_ambient` function in `util-linux`, and the loop condition is written like this: 
```c
for (cap = 0; cap < (sizeof(effective) * 8); cap++)
```
where `effective` is a `uint64_t`. Maybe we can also set the loop range to 0-63, but I don't quite understand why it's written as `(sizeof(effective) * 8)` here, because `effective` is declared as a 64-bit unsigned integer, so definitely `(sizeof(effective) * 8) == 64`.

Reply:
We should use
```python
    for cap in range(cap_last_cap + 1):
        if effective & (1 << cap):
            libc.prctl(libc.PR_CAP_AMBIENT, libc.PR_CAP_AMBIENT_RAISE, cap, 0, 0)
```
instead.

Reply:
> I don't quite understand why it's written as `(sizeof(effective) * 8)` here, because `effective` is declared as a 64-bit unsigned integer, so definitely `(sizeof(effective) * 8) == 64`.

It is good practice to not hardcode assumptions like `sizeof(effective) == 8` because it could change later on during a refactoring if someone changes the type of `effective`.